Builder API for the RoadGeometry

agalbachicar opened the pull request:
# 🎉 New feature

Part of #10 

## Summary

Opening this PR to discuss the Builder API of #10 

Scope: I don't expect to merge this PR as is but it should help to define the Builder API. The first two commits could be moved to independent PRs and make the last commit to work on top of the other two. We also need to define the LaneGeometry concept and class if required.

## Test it

Via unit tests

## Out of the scope of this PR

- BranchPoints
- LaneGeometry

agalbachicar commented on test/builder/builder_test.cc:
```diff
@@ -0,0 +1,122 @@
+
+  RoadGeometryBuilder dut;
+
+  // clang-format off
+  std::unique_ptr<maliput::api::RoadGeometry> rg =
+      dut.Id(kRoadGeometryId)
+          .LinearTolerance(kLinearTolerance)
+          .AngularTolerance(kAngularTolerance)
+          .ScaleLength(kScaleLength)
+          .InertialToBackendFrameTranslation(kInertialToBackendFrameTranslation)
+          .StartJunction()
+              .Id(kJunctionId)
+              .StartSegment()
+                      .Id(kSegmentAId)
+                  .StartLane()
+                      .Id(kLaneAId)
+                  .EndLane()
+                  .StartLane()
+                      .Id(kLaneBId)
+                  .EndLane()
+              .EndSegment()
+              .StartSegment()
+                  .Id(kSegmentBId)
+                  .StartLane()
+                      .Id(kLaneCId)
+                  .EndLane()
+              .EndSegment()
+          .EndJunction()
+          .Build();
+  // clang-format on
```
What do you think about the following @francocipollone @stonier ?

The geometric information of each Lane has not been added yet but I have thought of 2 options:

1- Create all the LineString3d somwhere else and then attach those to each Lane as a property like ID.
2- Create a LaneGeometryBuilder and expose that extra Builder and let the LaneBuilder be agnostic of the geometric details.

Thoughts?

francocipollone replied:
> 1- Create all the LineString3d somwhere else and then attach those to each Lane as a property like ID.

We could be benefit of this approach for reusing line strings: --> Linestring could be the left boundaries for one a lane and the right boundary for another one.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,219 @@
+  explicit SegmentBuilder(JunctionBuilder* parent) : details::NestedBuilder<JunctionBuilder>(parent) {}
+
+  SegmentBuilder& Id(const maliput::api::SegmentId& segment_id);
+  LaneBuilder StartLane();
```
How is it expected the lanes to be added? Related to the ordering of the lanes. Should the builder be smart enough in order to compute the ordering accordingly to the lane strings?

agalbachicar replied:
Good question.

I think the 2 target maps we have in mind now have a clear sense of left to right ordering, so we could let the parser derive that who has more understanding and context instead of the builder which needs to rely on lane specific information.

Alternatively, we could change this behavior by adding a sorting policy to the SegmentBuilder that is set before calling EndSegment(). That will make the EndSegment() function to sort lanes in a specific way.

francocipollone replied:
> I think the 2 target maps we have in mind now have a clear sense of left to right ordering, so we could let the parser derive that who has more understanding and context instead of the builder which needs to rely on lane specific information.

+1

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,219 @@
+template <typename ParentT>
+class NestedBuilder : public BuilderBase {
+ public:
+  explicit NestedBuilder(ParentT* parent) : parent_(parent) { MALIPUT_THROW_UNLESS(parent_ != nullptr); }
```
I find this nested builder convenient and natural to use at some point but I have some concerns when for example thinking downstream for the backends.

In order to go down in the hierarchy this api requires for me to know, pretty decently, how is the road geometry composed in a maliput-like fashion. Namely, to add a lane I need to know in which segment I will add it and in which junction I will add the segment.

From a user perspective I am not that confident that the user will be able to know this when parsing a file with a X format for road describing being loaded. Maybe, being able to load/build a lane and later on deciding which segment it should be added to, could be necessary for this situation? Wdyt @agalbachicar ?

agalbachicar replied:
I think that the scope of this API should be to only build a RoadGeometry. Any translation between the specific map format and the maliput graph must be implemented in the parser. With that responsibility separation in mind, we can change the API of the Builder to provider certain guarantees about what we will do with the provided information or favor one use or the other, but in the end, I think the closer to the RoadGeometry graph the better.

One thing that is a bit obscure and perhaps not too obvious is how BranchPoints will be created. I plan to work a bit more on that later by enabling a policy in the RoadGeometryBuilder to create BranchPoints:
- by LaneEnd overlap / intersection.
- by manually creating LaneEnds and LaneEndSets.

This is another relevant aspect of the connecitivity graph that is not contemplated for the moment.

francocipollone replied:
> Any translation between the specific map format and the maliput graph must be implemented in the parser

SG.

> One thing that is a bit obscure and perhaps not too obvious is how BranchPoints will be created.

>- by LaneEnd overlap / intersection.

For this non-manual case: Lanelet2 solves the connectivity by taking a look at the points that are shared among lanes. Two points defining an end of a lane and a start on another lane define a connection. We haven't defined having this kind of sharable data (points/linesstrings) but it is worth mentioning.

agalbachicar replied:
> For this non-manual case: Lanelet2 solves the connectivity by taking a look at the points that are shared among lanes. Two points defining an end of a lane and a start on another lane define a connection. We haven't defined having this kind of sharable data (points/linesstrings) but it is worth mentioning.

This is an interesting idea... I guess we could have some sort of associative container to index points and lanes. The key thing here is that so far we always linked centerline from Lane X with centerline from Lane Y. We might need to expose that information in the LaneGeometry to accomplish that before everything is created.

francocipollone reviewed (commented):
Thanks for pushing this front. 

I left some comments for discussion. I think that defining the api will provide us a guidance on the implementation under the hood.

I wonder if the user should have to know as much of the maliput's graph hierarchy (junction/segment/lane) or if the builder should be smart enough to identify adjacent lanes (segment) and connections (branchpoints) out of the line strings and points that are created.

agalbachicar commented:
@francocipollone let me know how you'd like me to split this PR. I know it is big.

francocipollone commented:
> @francocipollone let me know how you'd like me to split this PR. I know it is big.

This PR adds:
- Lane --> This can be removed, note that I opened a draft PR for adding Lane https://github.com/maliput/maliput_sparse/pull/16
- RoadGeometry --> A separate PR for presenting RoadGeometry class makes sense.
- BuilderApi --> This PR could be focused on this part.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+  LaneBuilder StartLane();
+
+  /**
+   * @brief Finanlizes the construction process of this Segment by inserting the Segment into the
```
nit: Typo Finanlizes

agalbachicar replied:
Done.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+ *     .Id("j0")
+ *     .StartSegment()
+ *       .Id("j0_s0")
+ *       .StartLane()
```
Wondering if we should provide a StartLeftLane and StartRightLane. In a way to suggest to the segment how the graph would be created independently of the geometry of the lanes.

agalbachicar replied:
I don't think we should. Each new line will be indexed from 0 to +inf and following the maliput::api::Segment::index() docstring, that means the order should be set in increasing +r direction. I created extra details in the SegmentBuilder::StartLane() method to clarify this point.

francocipollone reviewed (commented):
PTAL

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+ * @details Its sole purpose is to offer a template
+ * hierachy to avoid duplicating code that is related to storing the parent, returning
+ * it and making sure each children can call Passkey-protected methods.
+ */
```
Done. Good point.

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+  LaneGeometryBuilder StartLaneGeometry();
+
+  /**
+   * @brief Finanlizes the construction process of this Lane by inserting the Lane into the
```
Done.

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+  // void SetLaneGeometry(maliput::common::Passkey<LaneGeometryBuilder>, std::unique_ptr<LaneGeometry> lane_geometry);
+
+ private:
+  maliput::api::LaneId id_{"l_id"};
```
Done for all ids.

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+/**
+ * @brief Builder class for maliput::api::RoadGeometry.
+ */
+class RoadGeometryBuilder final : public details::BuilderBase {
```
No need, removed.

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+
+  /**
+   * @brief Sets the initial to backend frame translation vector of the maliput::api::RoadGeometry.
+   * @param translation The nitial to backend frame translation vector of the maliput::api::RoadGeometry.
```
Done.

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+   * @brief Builds a maliput::api::RoadGeometry.
+   * @details The underlying type of the RoadGeometry is maliput_sparse::RoadGeometry which is derived from
+   * maliput::geometry_base::RoadGeometry.
+   * @throws maliput::common::assertion_error  When there is no Junction to add to the RoadGeometry.
```
Done.

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,387 @@
+};
+
+}  // namespace builder
+}  // namespace maliput_sparse
```
Done.

agalbachicar commented on src/builder/builder.cc:
```diff
@@ -0,0 +1,152 @@
+  return End();
+}
+
+// TODO(maliput_sparse#10): Uncomment once LaneGeometry is ready.
```
Removed in favor of the proper implementation.

agalbachicar commented on src/builder/builder.cc:
```diff
@@ -0,0 +1,152 @@
+}
+
+}  // namespace builder
+}  // namespace maliput_sparse
```
Done.

agalbachicar commented:
So, @francocipollone I removed code from this PR and moved it to two different PRs. Then, because of requiring the implementation of the LaneGeometryBuilder I need to add tests and more code. In the end, this PR is longer than before but I honestly believe that is worth reviewing.

- There are ~100 LOC because of license notes.
- There are ~70 LOC because of cmake configuration
- The test file is long in lines of code but it is due to formatting. Similarly, the builder.h

Let me know if you want me to still split it.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+  LaneGeometryBuilder& LeftLineString(const maliput_sparse::geometry::LineString3d& left_line_string);
+
+  /// @brief Set the right maliput_sparse::geometry::LineString of the LaneGeometry.
+  /// @param left_line_string The right maliput_sparse::geometry::LineString to set in the LaneGeometry.
```
```suggestion
  /// @param right_line_string The right maliput_sparse::geometry::LineString to set in the LaneGeometry.
```

agalbachicar replied:
Done.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+
+ private:
+  maliput::api::LaneId id_{"unset_id"};
+  maliput::api::HBounds hbounds_{};
```
Probably we should add a default HBounds here. Like `{0., 5.}`

agalbachicar replied:
Done. I also changed some test cases to check the default value.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+
+  /// @brief Sets a maliput::geometry_base::Lane into this builder to fill in the Segment.
+  /// @details This method is only intended to be called by LaneBuilder instances. Call this method
+  /// in order to determine a specific left to right ordering for this Segment.
```
To avoid any misunderstanding, isn't it `right to left`?

agalbachicar replied:
Done.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+
+ private:
+  maliput::api::SegmentId id_{"unset_id"};
+  std::vector<std::unique_ptr<maliput::geometry_base::Lane>> lanes_{};
```
Use `maliput_sparse::Lane` instead

agalbachicar replied:
I did not want to include maliput::Lane because we need to expose the type and the SegmentBuilder actually does not care about the maliput_sparse::Lane but maliput::geometry_base::Lane.

francocipollone reviewed (commented):
The LOC is ok! Thanks for splitting the PR already.
LGTM I left some comments tho, ptal

agalbachicar commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+/// std::unique_ptr<RoadGeometry> road_geometry = RoadGeometryBuilder()
+///   .Id("two_lane_dragway")
+///   .StartJunction()
+///     .Id("j0")
```
I agree with the sentiment... I can change the default value to be initialized in the constructor with a different ID so there is no repetition. Most of the elements are required, meaning that you need to set them, and they can be configured in the same function most probably. But, because of splitting the process of constructing the entities I decided to be consistent and set each parameter via an independent function.

What do you think?

francocipollone replied:
Makes sense then :+1:

agalbachicar replied:
Note I did change some test cases in favor of testing default value too.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+
+/// @brief Base class of the builder's hierarchy.
+/// @details Its sole purpose is to offer a template
+/// hierachy to avoid duplicating code that is related to storing the parent, returning
+/// it and making sure each children can call Passkey-protected methods.
```
Sorry, I don't see what is BuilderBase class doing here. I think it is no longer needed. Could you enlighten me?

agalbachicar replied:
Nothing now, good catch!

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+
+/// @brief Holds the parent Builder class and offers a small set of convenient methods to manage
+/// the Builder lifecycle.
+/// @tparam ParentT The parent BuilderBase type.
```
Isn't this param to be also a NestedBuilder?

agalbachicar replied:
Not really. See `JunctionBuilder` for example.

francocipollone commented on include/maliput_sparse/builder/builder.h:
```diff
@@ -0,0 +1,360 @@
+ public:
+  /// @brief Construct a new Lane Geometry Builder object.
+  /// @param parent The parent LaneBuilder. It must not be nullptr.
+  explicit LaneGeometryBuilder(LaneBuilder* parent) : details::NestedBuilder<LaneBuilder>(parent) {}
```
Do we miss an assertion here? Same for all the constructors

agalbachicar replied:
Not really. `NestedBuilder` constructor is in charge of that.

francocipollone reviewed (approved):
Great Job! Left some nits

francocipollone reviewed (approved):
LGTM!
